dkg/actions_active.go

```diff
@@ -0,0 +1,443 @@
+
+	me, err := d.identityForBeacon(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	lastDKG, err := d.store.GetFinished(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	newEpoch := lastDKG.Epoch + 1
+	terms := drand.ProposalTerms{
+		BeaconID:             beaconID,
+		Threshold:            options.Threshold,
+		Epoch:                newEpoch,
+		SchemeID:             lastDKG.SchemeID,
+		BeaconPeriodSeconds:  uint32(lastDKG.BeaconPeriod.Seconds()),
+		CatchupPeriodSeconds: options.CatchupPeriodSeconds,
+		GenesisTime:          timestamppb.New(lastDKG.GenesisTime),
+		GenesisSeed:          lastDKG.GenesisSeed,
+		TransitionTime:       options.TransitionTime,
+		Timeout:              options.Timeout,
+		Leader:               me,
+		Joining:              options.Joining,
+		Remaining:            options.Remaining,
+		Leaving:              options.Leaving,
+	}
+
+	nextState, err := lastDKG.Proposing(me, &terms)
+	if err != nil {
+		return nil, err
+	}
+
+	sendProposalToAllAndStoreState := func() error {
+		// we send the proposal to joiners and remainers and error if they don't respond
+		err = d.network.Send(
+			me,
+			util.Concat(nextState.Joining, nextState.Remaining),
+			func(client net.DKGClient, peer net.Peer) (*drand.EmptyResponse, error) {
+				return client.Propose(ctx, peer, &terms)
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+		// we make a best effort attempt to send the proposal to the leaver, but if their node is e.g. turned off then
+		// we ignore the error
+		if len(nextState.Leaving) > 0 {
+			err = d.network.Send(me, nextState.Leaving, func(client net.DKGClient, peer net.Peer) (*drand.EmptyResponse, error) {
+				return client.Propose(ctx, peer, &terms)
+			})
+
+			if err != nil {
+				d.log.Warnw("could not send proposal to a leaving participant", "err", err)
+			}
+		}
+
+		if err := d.store.SaveCurrent(beaconID, nextState); err != nil {
+			return err
+		}
+
+		d.log.Infow("Finished proposing a new DKG", "beaconID", beaconID)
+		return nil
+	}
+
+	// if there's an error sending to a party or saving the state, attempt a rollback by issuing an abort
+	rollback := func(err error) {
+		allParticipants := util.Concat(nextState.Joining, nextState.Remaining, nextState.Leaving)
+		d.log.Errorw("There was an error proposing a DKG", "err", err, "beaconID", beaconID)
+		_ = d.attemptAbort(ctx, me, allParticipants, beaconID)
+	}
+
+	return responseOrError(rollbackOnError(sendProposalToAllAndStoreState, rollback))
+}
+
+func (d *DKGProcess) StartAbort(ctx context.Context, options *drand.AbortOptions) (*drand.EmptyResponse, error) {
+	beaconID := options.BeaconID
+	d.log.Infow("Aborting DKG", "beaconID", beaconID)
+
+	me, err := d.identityForBeacon(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	current, err := d.store.GetCurrent(beaconID)
+	if err != nil {
+		return nil, err
+	}
+
+	if !util.EqualParticipant(current.Leader, me) {
+		return nil, errors.New("cannot abort the DKG if you aren't the leader")
```

Comment:
really ? Im confused. 
Once the DKG is launched, or even before, any one should be able to take the decision to abort if one rule or something they see don't correspond.